experiment/popularity_community/crawl_torrents.py

```diff
@@ -0,0 +1,217 @@
+from ipv8.peer import Peer
+from ipv8.peerdiscovery.discovery import RandomWalk
+from tool.tiny_tribler_service import TinyTriblerService
+from tribler_core.modules.metadata_store.community.remote_query_community import RemoteQueryCommunity, \
+    SelectResponsePayload, SelectRequest, RemoteSelectPayload
+from tribler_core.modules.metadata_store.serialization import REGULAR_TORRENT
+from tribler_core.utilities.unicode import hexlify
+
+UNLIMITED = -1  # this magic number is use inside many Tribler modules
+
+IPV8_WALK_INTERVAL = 0.05
+
+RANDOM_WALK_TIMEOUT_IN_SEC = 5
+RANDOM_WALK_WINDOW_SIZE = 0
+RANDOM_WALK_RESET_CHANCE = 30
+
+CRAWLER_REQUEST_TIMEOUT_IN_SEC = 60
+CRAWLER_SAVE_INTERVAL_IN_SEC = 60
+
+
+class CrawlerSelectRequest(SelectRequest):
+    """Custom class for RemoteQueryCommunity select request
+
+    It has been added only for having an ability to increase `timeout_delay`
+    """
+
+    @property
+    def timeout_delay(self):
+        return CRAWLER_REQUEST_TIMEOUT_IN_SEC
+
+
+class Crawler(RemoteQueryCommunity):
+    """ Crawler based on RemoteQueryCommunity
+
+    The approach based on two ideas:
+        * use "aggressive" values for network walking (see ExperimentConfiguration.RANDOM_WALK_*)
+        * use RemoteQueryCommunity's ability to select MetadataInfo from remote db
+    """
+
+    def __init__(self, my_peer, endpoint, network, metadata_store, out_file_path):
+        super().__init__(my_peer, endpoint, network, metadata_store)
+        self._logger = logging.getLogger(self.__class__.__name__)
+        self._output_file_path = out_file_path
+        self.max_peers = UNLIMITED
+        self.queried_peers_limit = UNLIMITED
+
+        self.torrent_dictionary = defaultdict(list)  # {"Peer": (votes, infohash, title)}
+        self.polled_peers = set()
+
+        self.register_task("request_torrents", self.request_torrents_from_new_peers, interval=10)
+        self.register_task("save_torrents", self.save, interval=CRAWLER_SAVE_INTERVAL_IN_SEC)
+
+    def __del__(self):
+        self.save()
+
+    def request_torrents_from_new_peers(self):
+        """Send a request for gathering torrent information
+
+        Behaviour:
+
+        For all new pears: ```self.get_peers() - self.polled_peers```
+        the crawler request 100 most popular torrents.
+
+        Note:   it is possible to request more, but it requires a significant
+                increase of code complexity.
+        Ref: https://github.com/Tribler/tribler/blob/devel/src/tribler-core/tribler_core/modules/metadata_store/community/remote_query_community.py#L19
+        """
+        peers_ready_to_request = set(self.get_peers()) - self.polled_peers
+        self._logger.info(f"Request torrents from {peers_ready_to_request}")
+
+        for peer in peers_ready_to_request:
+            self._logger.debug(f"Requesting torrents from peer: {peer}")
+            self.send_select(peer, metadata_type=[REGULAR_TORRENT], first=0, last=100)
```

Comment:
Ah! How could I miss that?
You're not **sorting** the torrents on health! Therefore, you will get essentially semi-random torrents. To enable sorting, you must add `sort_by=HEALTH` to the list of `send_select` arguments.

Reply:
Also, sorting on health makes the *order* of torrents in the packets important. You can write that down to establish a partial order on the set of torrents. Later you can compare and combine partial orders from different peers.

Reply:
Fixed